pkg/controller/cronjob/metrics/metrics.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const CronJobControllerSubsystem = "cronjob_controller"
+
+var (
+	CronJobCreationSkew = metrics.NewHistogram(
+		&metrics.HistogramOpts{
+			Subsystem:      CronJobControllerSubsystem,
+			Name:           "cronjob_job_creation_skew_duration_seconds",
+			Help:           "Time between when a cronjob is scheduled to be run, and when the corresponding job is created",
+			StabilityLevel: metrics.ALPHA,
+			Buckets:        []float64{1, 2.5, 5, 10, 15, 25, 50, 120, 300, 600, 1200},
+		},
+	)
+	CronJobMissedSchedulesTotal = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Subsystem:      CronJobControllerSubsystem,
+			Name:           "cronjob_missed_schedules",
+			Help:           "Number of schedules missed for a cronjob before a job successfully was created",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"namespace", "name"},
```

Comment:
These are useful labels, but will create a memory leak in the controller, since label combinations are only ever added (with `.Set(` below), and aren't ever removed.  You can either:

1. Remove all unbounded dimensions (namespace + name), or
2. Implement the [`StableCollector` interface](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/collector.go#L28), which allows you to determine at scrape time which metric streams to omit.  This would allow you to "drop" unwanted streams by not emitting them.

Reply:
I think the namespace and name are important metadata for the missed schedules metric to make sense, because from the namespace and name a user can look at what the `spec.schedule` value is and understand how severe the missed schedules are. For example, 5 missed schedules for a corn job scheduled every hour and 5 missed schedules for a cron scheduled every week has a very different severity level.

Since there is non-trivial work involved in adding this metric, do you mind if I take adding this metric as a follow-up? @dashpole

Reply:
I'm just here to make sure you don't add memory leaks :).  Its up to the reviewers/approvers of the controller code which metrics are required for the PR to be merged.  I'm happy to review separate PRs for each metric.